cmd/libsnap-confine-private/device-cgroup-support.c

```diff
@@ -107,6 +111,322 @@ static void _sc_cgroup_v1_attach_pid(sc_device_cgroup *self, pid_t pid) {
+        BPF_JMP_IMM(BPF_JEQ, BPF_REG_0, 0, 2),                               /* if (value_ptr == 0) goto pc + 2 */
+        /* we found a match with any minor number for that type|major */
+        BPF_MOV64_IMM(BPF_REG_0, 1), /* r0 = 1 */
+        BPF_JMP_A(1),
+        BPF_MOV64_IMM(BPF_REG_0, 0), /* r0 = 0 */
+        BPF_EXIT_INSN(),
+    };
+
+    char log_buf[4096] = {0};
+
+    int prog_fd =
+        bpf_load_prog(BPF_PROG_TYPE_CGROUP_DEVICE, prog, sizeof(prog) / sizeof(prog[0]), log_buf, sizeof(log_buf));
+    if (prog_fd < 0) {
+        die("cannot load program:\n%s\n", log_buf);
+    }
+    return prog_fd;
+}
+
+static int _sc_cgroup_v2_init_bpf(sc_device_cgroup *self, int flags) {
+    self->v2.devmap_fd = -1;
+    self->v2.cgroup_fd = -1;
+
+    char *own_group SC_CLEANUP(sc_cleanup_string) = sc_cgroup_v2_own_path_full();
+    if (own_group == NULL) {
+        die("cannot obtain own group path");
+    }
+
+    char own_group_full[PATH_MAX] = {0};
+    sc_must_snprintf(own_group_full, sizeof(own_group_full), "/sys/fs/cgroup/%s", own_group);
+    int cgroup_fd = open(own_group_full, O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+    if (cgroup_fd < 0) {
+        die("cannot open own cgroup directory %s", own_group_full);
+    }
+    debug("cgroup %s opened at %d", own_group_full, cgroup_fd);
+
+    self->v2.tag = sc_strdup(self->security_tag);
+    /* bpffs is unhappy about dots in the name, replace all with underscores */
+    for (char *c = strchr(self->v2.tag, '.'); c != NULL; c = strchr(c, '.')) {
+        *c = '_';
+    }
+
+    char path[PATH_MAX] = {0};
+    sc_must_snprintf(path, sizeof path, "/sys/fs/bpf/snap/%s", self->v2.tag);
+
+    /* TODO: open("/sys/fs/bpf") and then mkdirat?  */
+    sc_identity old = sc_set_effective_identity(sc_root_group_identity());
+    if (mkdir("/sys/fs/bpf/snap", 0700) < 0) {
+        if (errno != EEXIST) {
+            die("cannot create /sys/fs/bpf/snap directory");
+        }
+    }
+    (void)sc_set_effective_identity(old);
+
+    /* XXX: this should be more than enough keys */
+    const size_t max_entries = 500;
+    int devmap_fd = bpf_get_by_path(path);
+    if (devmap_fd < 0) {
+        if (errno != ENOENT) {
+            die("cannot attempt to get existing device map");
+        }
+        if (flags == SC_DEVICE_CGROUP_FROM_EXISTING) {
+            /* there is no map, and we haven't been asked to setup a new cgroup */
+            return -1;
+        }
+        debug("device map not present yet");
+        /* map not created and pinned yet */
+        const size_t value_size = 1;
+        devmap_fd = bpf_create_map(BPF_MAP_TYPE_HASH, sizeof(struct sc_cgroup_v2_device_key), value_size, max_entries);
+        if (devmap_fd < 0) {
+            die("cannot create bpf map");
+        }
+        debug("got bpf map at fd: %d", devmap_fd);
+        sc_identity old = sc_set_effective_identity(sc_root_group_identity());
```

Comment:
also some comments about why/when we need to be root for some bpf operations but not others would be appreciated, for example why is it okay that we create the map as non-root, but in order to pin it to a path we need to be root? A simple "this is what the docs say" or something would be helpful, but I'm nervous about approving code which switches uids around without understanding why or when we are supposed to be doing so.

Reply:
Added some comments

Reply:
Sorry I can't find any comments re switching group id - can you point me to where they are..?

Reply:
Heh, just noticed that the comments and fixes requested in earlier reviews were not pushed to this branch, but instead to #10575. I'll push them here right away.